File: src/utils/dateFormatter.test.ts
import { describe, it, expect, vi, afterEach } from 'vitest';
import { formatChatTime, formatConversationTime } from './dateFormatter';

describe('dateFormatter', () => {
  afterEach(() => {
    vi.useRealTimers();
  });

  describe('formatChatTime', () => {
    it('should show time only for today', () => {
      const now = new Date();
      now.setHours(14, 30, 0, 0);
      vi.setSystemTime(now);

      const result = formatChatTime(now.toISOString());
      expect(result).toBe('14:30');
    });

    it('should show "昨天" for yesterday', () => {
      vi.setSystemTime(new Date('2024-06-15T10:00:00'));
      const yesterday = new Date('2024-06-14T09:15:00');
      const result = formatChatTime(yesterday.toISOString());
      expect(result).toBe('昨天 09:15');
    });

    it('should show weekday for dates within 6 days', () => {
      vi.setSystemTime(new Date('2024-06-15T10:00:00')); // Saturday
      const threeDaysAgo = new Date('2024-06-12T08:00:00'); // Wednesday
      const result = formatChatTime(threeDaysAgo.toISOString());
      expect(result).toBe('周三 08:00');
    });

    it('should show MM/DD for older dates', () => {
      vi.setSystemTime(new Date('2024-06-15T10:00:00'));
      const oldDate = new Date('2024-05-01T14:00:00');
      const result = formatChatTime(oldDate.toISOString());
      expect(result).toBe('05/01 14:00');
    });
  });

  describe('formatConversationTime', () => {
    it('should show time for today', () => {
      const now = new Date();
      now.setHours(9, 5, 0, 0);
      vi.setSystemTime(now);

      const result = formatConversationTime(now.toISOString());
      expect(result).toBe('09:05');
    });

    it('should show "昨天" for yesterday', () => {
      vi.setSystemTime(new Date('2024-06-15T10:00:00'));
      const yesterday = new Date('2024-06-14T09:15:00');
      const result = formatConversationTime(yesterday.toISOString());
      expect(result).toBe('昨天');
    });

    it('should show weekday within 6 days', () => {
      vi.setSystemTime(new Date('2024-06-15T10:00:00')); // Saturday
      const date = new Date('2024-06-12T08:00:00'); // Wednesday
      const result = formatConversationTime(date.toISOString());
      expect(result).toBe('周三');
    });

    it('should show MM/DD for older dates', () => {
      vi.setSystemTime(new Date('2024-06-15T10:00:00'));
      const oldDate = new Date('2024-01-15T14:00:00');
      const result = formatConversationTime(oldDate.toISOString());
      expect(result).toBe('01/15');
    });
  });
});

File: lang_buddy/test/unit/core/date_formatter_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:lang_buddy/core/utils/date_formatter.dart';

void main() {
  group('DateFormatter.formatChatTime', () {
    test('returns HH:mm for today', () {
      final now = DateTime.now();
      final today = DateTime(now.year, now.month, now.day, 14, 30);
      final result = DateFormatter.formatChatTime(today);
      expect(result, '14:30');
    });

    test('returns 昨天 HH:mm for yesterday', () {
      final yesterday = DateTime.now().subtract(const Duration(days: 1));
      final dt = DateTime(yesterday.year, yesterday.month, yesterday.day, 9, 15);
      final result = DateFormatter.formatChatTime(dt);
      expect(result, startsWith('昨天'));
      expect(result, contains('09:15'));
    });

    test('returns weekday name for 2-6 days ago', () {
      final twoDaysAgo = DateTime.now().subtract(const Duration(days: 2));
      final dt = DateTime(
          twoDaysAgo.year, twoDaysAgo.month, twoDaysAgo.day, 10, 0);
      final result = DateFormatter.formatChatTime(dt);
      expect(
        result,
        anyOf([
          '周一', '周二', '周三', '周四', '周五', '周六', '周日',
        ]),
      );
    });

    test('returns MM/dd for older messages', () {
      final oldDate = DateTime.now().subtract(const Duration(days: 30));
      final dt = DateTime(oldDate.year, oldDate.month, oldDate.day, 8, 0);
      final result = DateFormatter.formatChatTime(dt);
      expect(result, matches(RegExp(r'\d{2}/\d{2}')));
    });
  });

  group('DateFormatter.formatConversationTime', () {
    test('returns HH:mm for today', () {
      final now = DateTime.now();
      final today = DateTime(now.year, now.month, now.day, 18, 45);
      final result = DateFormatter.formatConversationTime(today);
      expect(result, '18:45');
    });

    test('returns 昨天 for yesterday', () {
      final yesterday = DateTime.now().subtract(const Duration(days: 1));
      final dt = DateTime(yesterday.year, yesterday.month, yesterday.day, 10, 0);
      final result = DateFormatter.formatConversationTime(dt);
      expect(result, '昨天');
    });

    test('returns weekday name for 2-6 days ago', () {
      final threeDaysAgo = DateTime.now().subtract(const Duration(days: 3));
      final dt = DateTime(
          threeDaysAgo.year, threeDaysAgo.month, threeDaysAgo.day, 10, 0);
      final result = DateFormatter.formatConversationTime(dt);
      expect(
        result,
        anyOf([
          '周一', '周二', '周三', '周四', '周五', '周六', '周日',
        ]),
      );
    });

    test('returns MM/dd for older conversations', () {
      final oldDate = DateTime.now().subtract(const Duration(days: 14));
      final dt = DateTime(oldDate.year, oldDate.month, oldDate.day, 12, 0);
      final result = DateFormatter.formatConversationTime(dt);
      expect(result, matches(RegExp(r'\d{2}/\d{2}')));
    });

    test('conversation time is shorter than chat time for yesterday', () {
      final yesterday = DateTime.now().subtract(const Duration(days: 1));
      final dt = DateTime(yesterday.year, yesterday.month, yesterday.day, 14, 30);
      final chatTime = DateFormatter.formatChatTime(dt);
      final convTime = DateFormatter.formatConversationTime(dt);
      // Chat time includes HH:mm, conversation time just says 昨天
      expect(chatTime.length, greaterThan(convTime.length));
    });
  });
}
